docs/04-contributing/development_environment.md

```diff
@@ -0,0 +1,239 @@
+- for production there are builder targets to build the frontend and backend of mbin
+- for development, the `dev` target is minimal as the frontend and backend are built when the services are first run 
+- the final target is `prod` which unites the outputs of the builder targets
+
+### Building the frontend and backend
+
+As mentioned [above](#phpdockerfile), the front and backend are built in the `Dockerfile` for production.
+
+In development, this is taken care of at runtime in [compose.dev.yml]. The `messenger` service is the first PHP
+service to run, so it installs PHP dependencies and such. The `node` service is only in the dev compose not production
+and requires files from the PHP dependencies to successfully build the frontend. It is thus run afterward.
+
+
+## Frequently executed tasks
+
+### Modifying behavior with a quick turn-around
+
+The entire repository is mounted into the `caddy`, `php`, and `messenger` services.
+
+**PHP code**
+
+Most PHP code changes are picked up immediately after refreshing the page in the browser.
+
+**Other important files**
+
+| file          | services       | how to refresh                                                         |
+|---------------|----------------|------------------------------------------------------------------------|
+| Caddyfile     | caddy          | `docker compose exec caddy caddy reload --config /etc/caddy/Caddyfile` |
+| entrypoint.sh | php, messenger | `docker compose up -d --no-deps $service`                              |
+
+
+### Debugging PHP (XDebug)
+
+PHP supports remote debugging using [XDebug](https://xdebug.org/). This works by hosting an Xdebug server the php
+process can call. 
+
+```mermaid
+sequenceDiagram
+    participant Browser
+    participant rp as Reverse Proxy
+    participant php as PHP-FPM
+    participant xdebug as XDebug Host
+
+    Browser->>+rp: http://localhost
+    rp->>+php: index.php
+    php->>+xdebug: :9003
+    xdebug->>php: break
+    php-->>xdebug: 
+    xdebug->>php: continue
+    php-->>xdebug: 
+    xdebug-->>-php: done
+    php-->>-rp: response
+    rp-->>-Browser: response
+```
+
+**Requirement**
+
+An XDebug server. The XDebug server is often hosted on port 9000 or 9003. Some IDEs (like [PHPStorm]) have it builtin.
+
+**Enabling**
+
+Open [app.dev.ini] and uncomment `;xdebug.start_with_request=yes` by removing the `;`, 
+then restart the `php` service using `docker compose restart php`.
+
+Once you navigate to a page, your IDE/editor should get called from `php-fpm` within docker.
+
+#### Special cases
+
+Firewalls can sometimes get in the way of communication between the docker container and the docker host.
+
+**NixOS**
+
+Nixos needs [iptables rules](https://discourse.nixos.org/t/docker-container-not-resolving-to-host/30259/8).
+
+# Development Server
+
+Requirements:
+
+- PHP v8.2
+- NodeJS
+- Redis
+- PostgreSQL
+- _Optionally:_ Mercure
+
+---
+
+- Increase execution time in PHP config file: `/etc/php/8.2/fpm/php.ini`:
```

Comment:
Is this PR rebased correctly? Since it should be promoting PHP 8.3.

```suggestion
- Increase execution time in PHP config file: `/etc/php/8.3/fpm/php.ini`:
```